packages/core/src/middleware/koa-log.ts

```diff
@@ -1,21 +1,20 @@
-import { LogResult, LogType } from '@logto/schemas';
-import { Context, MiddlewareType } from 'koa';
+import { LogPayload, LogPayloads, LogResult, LogType } from '@logto/schemas';
+import { Optional } from '@silverhand/essentials';
+import deepmerge from 'deepmerge';
+import { MiddlewareType } from 'koa';
 import { nanoid } from 'nanoid';
 
 import { insertLog } from '@/queries/log';
 
 export type WithLogContext<ContextT> = ContextT & {
-  log: LogContext;
+  log: <T extends LogType, P extends LogPayloads[T]>(type: T, payload: P) => void;
 };
 
-export interface LogContext {
-  [key: string]: unknown;
-  type?: LogType;
-}
-
-const log = async (ctx: WithLogContext<Context>, result: LogResult) => {
-  const { type, ...rest } = ctx.log;
-
+const saveLog = async (
+  type: Optional<LogType>,
+  payload: Optional<LogPayload>,
+  result: LogResult
+) => {
```

Comment:
Optional function params should follow required ones, not the opposite.

Reply:
If the route doesn't set the log type, it will not save log content.

Refactored. PTAL